packages/next/src/build/utils.ts

```diff
@@ -1133,100 +1124,152 @@ export type AppConfigDynamic =
   segmentPath: string
   getStaticPaths?: GetStaticPaths
-  generateStaticParams?: any
+  generateStaticParams?: GenerateStaticParams
   isLayout?: boolean
-}>
+}
+
+export type GenerateParams = Array<SegmentGenerateParams>
 
-export const collectAppConfig = (mod: any): AppConfig | undefined => {
+function collectAppConfig(mod: object): AppConfig | undefined {
   let hasConfig = false
 
   const config: AppConfig = {}
-  if (typeof mod?.revalidate !== 'undefined') {
+  if (
+    'revalidate' in mod &&
+    (typeof mod.revalidate === 'number' || mod.revalidate === false)
+  ) {
     config.revalidate = mod.revalidate
     hasConfig = true
   }
-  if (typeof mod?.dynamicParams !== 'undefined') {
+
+  if ('dynamicParams' in mod && typeof mod.dynamicParams === 'boolean') {
     config.dynamicParams = mod.dynamicParams
     hasConfig = true
   }
-  if (typeof mod?.dynamic !== 'undefined') {
-    config.dynamic = mod.dynamic
+
+  // TODO: add config validation
+  if ('dynamic' in mod && typeof mod.dynamic === 'string') {
+    config.dynamic = mod.dynamic as AppConfigDynamic
     hasConfig = true
   }
-  if (typeof mod?.fetchCache !== 'undefined') {
-    config.fetchCache = mod.fetchCache
+
+  // TODO: add config validation
+  if ('fetchCache' in mod && typeof mod.fetchCache === 'string') {
+    config.fetchCache = mod.fetchCache as AppConfigFetchCache
     hasConfig = true
   }
-  if (typeof mod?.preferredRegion !== 'undefined') {
+
+  if ('preferredRegion' in mod && typeof mod.preferredRegion === 'string') {
     config.preferredRegion = mod.preferredRegion
     hasConfig = true
   }
 
   return hasConfig ? config : undefined
 }
 
-export const collectGenerateParams = async (
-  segment: any,
-  parentSegments: string[] = [],
-  generateParams: GenerateParams = []
-): Promise<GenerateParams> => {
-  if (!Array.isArray(segment)) return generateParams
-  const isLayout = !!segment[2]?.layout
-  const mod = await (isLayout
-    ? segment[2]?.layout?.[0]?.()
-    : segment[2]?.page?.[0]?.())
-  const config = collectAppConfig(mod)
-  const page: string | undefined = segment[0]
-  const isClientComponent = isClientReference(mod)
-  const isDynamicSegment = /^\[.+\]$/.test(page || '')
-  const { generateStaticParams, getStaticPaths } = mod || {}
-
-  //console.log({parentSegments, page, isDynamicSegment, isClientComponent, generateStaticParams})
-  if (isDynamicSegment && isClientComponent && generateStaticParams) {
-    throw new Error(
-      `Page "${page}" cannot export "generateStaticParams()" because it is a client component`
-    )
-  }
+/**
+ * Walks the loader tree and collects the generate parameters for each segment.
+ *
+ * @param tree the loader tree
+ * @returns the generate parameters for each segment
+ */
+export async function collectGenerateParams(tree: LoaderTree) {
+  const generateParams: GenerateParams = []
+  const parentSegments: string[] = []
+
+  const stack: LoaderTree[] = [tree]
+  while (stack.length > 0) {
```

Comment:
Seems like the stack is unnecessary, it only ever has one element in it. Just track this with a variablelike

```
let currentLoaderTree = tree
while(currentLoaderTree) {
  ...
  currentLoaderTree = parallelRoutes.children
}